qa/rpc-tests/rest.py

```diff
@@ -339,5 +339,112 @@ def run_test(self):
+        assert_equal(response.status, 200)
+        response_str = response.read()
+        assert_equal(self.nodes[0].sendrawtransaction(raw_tx['hex']) in response_str, True);
+
+        self.nodes[0].generate(1)
+        self.sync_all()
+        self.nodes[2].generate(1)
+        self.sync_all()
+
+        #########
+        # sendrawtransaction with transaction that is already in the block chain
+        #########
+        response = http_post_call(url.hostname, url.port, '/rest/tx'+self.FORMAT_SEPARATOR+"hex", "transaction=" + raw_tx['hex'] , True)
+        assert_equal(response.status, 400)
+        response_str = response.read()
+        assert_equal("Transaction already in block chain" in response_str, True);
+
+        #########
+        # sendrawtransaction as hex, requesting result as json
+        #########
+        txId = self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), 5)
+        inputs  = [ {'txid' : txId, 'vout' : 1}]
+        outputs = { self.nodes[0].getnewaddress() : 4.6 }
+        raw_tx   = self.nodes[0].createrawtransaction(inputs, outputs)
+        raw_tx   = self.nodes[0].signrawtransaction(raw_tx)
+        response = http_post_call(url.hostname, url.port, '/rest/tx'+self.FORMAT_SEPARATOR+"json", "transaction=" + raw_tx['hex'] , True)
+        response_str = response.read()
+        print response.status #lclc RAUS
+        print response_str #lclc RAUS
+        assert_equal(response.status, 200)
+        assert_equal( "{ \"txid\" : \"" +self.nodes[0].sendrawtransaction(raw_tx['hex']) + "\" }" in response_str, True);
+
+        self.nodes[0].generate(1)
+        self.sync_all()
+
+
+        #########
+        # sendrawtransaction as binary, requesting result as binary
+        #########
+        txId = self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), 5.0)
+        inputs  = [ {'txid' : txId, 'vout' : 1}]
+        outputs = { self.nodes[0].getnewaddress() : 4.4 }
+        raw_tx   = self.nodes[0].createrawtransaction(inputs, outputs)
+        raw_tx   = self.nodes[0].signrawtransaction(raw_tx)
+
+        binaryRequest = binascii.unhexlify(raw_tx['hex'])
+        print binaryRequest #lclc RAUS
+        print 1 #lclc RAUS
+        print "transaction="+binaryRequest #lclc RAUS
+        response = http_post_call(url.hostname, url.port, '/rest/tx'+self.FORMAT_SEPARATOR+"bin", "transaction="+binaryRequest , True)
+#        response = http_post_call(url.hostname, url.port, '/rest/tx'+self.FORMAT_SEPARATOR+"hex", "transaction=" + raw_tx['hex'] , True) #lclc RAUS
+
+        print response.status #lclc RAUS
+#        assert_equal(response.status, 200)
+        response_str = response.read()
+        print response_str #lclc RAUS
+        response_str = binascii.hexlify(response.read())
+        print response_str #lclc RAUS
+        print 2 #lclc RAUS
+        print self.nodes[0].sendrawtransaction(raw_tx['hex']) #lclc RAUS
```

Comment:
Lucas: in a raw API the creator of the transactions _cannot_ verify the
fee, because he doesn't have the actual network's txouts, so they don't
know the amounts being spent.

It's a serious layer violation IMHO to make AcceptToMemoryPool perform this
check, but I do agree that sendrawtransaction or its equivalent REST
interface should do this validation. People do shoot themselves in the foot
by spending large inputs and not setting a (correct) change, crazy as it
sounds.